FuelLabs/fuel-indexer
# Standardize fuel-indexer-api-server graphql endpoint

### Description
- This PR swaps `rustc` on CI back to `lastest` (1.63) from 1.62
- This PR updates `fuel-indexer-api-server` so that it's GraphQL endpoint mirrors the `fuel-indexer` GraphQL endpoint
  - Also adds a `run_migrations` GraphQL configuration variable (this could be added to the node in genera (e.g., run migrations on Postgres/GraphQL), but not really necessary at the moment)
  
### Testing steps
- The testing steps are a little involved
  - If you wanted to be thorough you could checkout `master` and see that the functionality is broken before testing the fix
- Steps:
  - [ ] Start the example's webapp `cd examples/simple-native/counter-rs && RUST_LOG=info cargo run`
  - [ ] Start the example's indexer service `cd examples/simple-native && RUST_LOG=info cargo run -- --fuel-node-port 4004 --postgres-database indexer --test-manifest manifest.yaml`
  - [ ] Bootstrap a database with some schema containing a FK constraint
    - We'll use the `simple-native` example to test

```sql
insert into counter.count values (1, 1, 1659465533, '1'::Bytea);
insert into counter.anothercount values (1, 1, 1659465535, '0x366f3b631cf475A13402b6A04cF226f86D7B7921', 1, '1'::Bytea);
insert into counter.anothercount values (2, 1, 1659465901, '0x444f3b631cf475A13402b6A04cF226f86D7B7921', 1, '1'::Bytea);
```
  - [ ] Start the api-server `RUST_LOG=debug cargo run --bin fuel-indexer-api-server -- --graphql-api-port 29988 --postgres-database indexer` (with your proper config options)
  - [ ] Query the `fuel-indexer` service in the `simple-native` example via `curl`. On this PR branch, your response should look as follows

```bash
➜  curl -X POST http://localhost:29987/graph/counter -H 'content-type: application/json' -d '{"query": "query { anothercount { id timestamp reference_count }}", "params": "b"}' | json_pp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   187  100   105  100    82    944    737 --:--:-- --:--:-- --:--:--  2200
[
   {
      "id" : 1,
      "reference_count" : 1,
      "timestamp" : 1659465535
   },
   {
      "id" : 2,
      "reference_count" : 1,
      "timestamp" : 1659465901
   }
]
```
  - [ ] Query the `fuel-indexer-api-server` service via `curl`. On this PR branch, your response should match the response you get from the `fuel-indexer` service above ☝️ 

```bash
➜  curl -X POST http://0.0.0.0:29987/graph/counter -H 'content-type: application/json' -d '{"query": "query { anothercount { id timestamp reference_count }}", "params": "b"}' | json_pp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   187  100   105  100    82    944    737 --:--:-- --:--:-- --:--:--  2200
[
   {
      "id" : 1,
      "reference_count" : 1,
      "timestamp" : 1659465535
   },
   {
      "id" : 2,
      "reference_count" : 1,
      "timestamp" : 1659465901
   }
]
```
  - [ ] If you make the same query on `master` your result should resemble (not that only one item in the on-to-many relationship was returned):
```bash
➜  curl -X POST http://0.0.0.0:29988/graph/counter -H 'content-type: application/json' -d '{"query": "query { anothercount { id timestamp reference_count }}", "params": "b"}' | json_pp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   133  100    51  100    82   3317   5333 --:--:-- --:--:-- --:--:-- 16625
{
   "id" : 1,
   "reference_count" : 1,
   "timestamp" : 1659465535
}
```
